src/main/java/io/confluent/kafkarest/ConsumerReadTask.java

```diff
@@ -108,37 +110,50 @@ public boolean doPartialRead() {
       long startedIteration = parent.getConfig().getTime().milliseconds();
       final int requestTimeoutMs = parent.getConfig().getInt(
           KafkaRestConfig.CONSUMER_REQUEST_TIMEOUT_MS_CONFIG);
+      final long endTime = startedIteration + requestTimeoutMs;
+      final int itBackoff = parent.getConfig().getInt(
+        KafkaRestConfig.CONSUMER_ITERATOR_BACKOFF_MS_CONFIG);
       try {
-        // Read off as many messages as we can without triggering a timeout exception. The
-        // consumer timeout should be set very small, so the expectation is that even in the
-        // worst case, num_messages * consumer_timeout << request_timeout, so it's safe to only
-        // check the elapsed time once this loop finishes.
-        while (iter.hasNext()) {
-          MessageAndMetadata<KafkaK, KafkaV> msg = iter.peek();
-          ConsumerRecordAndSize<ClientK, ClientV> recordAndSize = parent.createConsumerRecord(msg);
-          long roughMsgSize = recordAndSize.getSize();
-          if (bytesConsumed + roughMsgSize > maxResponseBytes) {
-            break;
+        while (parent.getConfig().getTime().milliseconds() < endTime) {
+
+          if (iter == null || !iter.hasNext()) {
+            // The consumer timeout should be set very small, so the expectation is that even in the
+            // worst case, num_messages * consumer_timeout << request_timeout, so it's safe to only
+            // check the elapsed time once this loop finishes.
+            ConsumerRecords<KafkaK, KafkaV> records = consumer.poll(itBackoff);
```

Comment:
Wouldn't it better to give each `ConsumerState` a blocking queue and to have a separate thread that periodically performs poll(0) on each `KafkaConsumer` object? Retrieved records are put into that queue. When a user asks consumer for records:
1) Records from queue are retrieved first and it becomes empty. If the queue is not empty go to #3
2) poll with some timeout is invoked (We can still use Round-Robin strategy to poll for records by invoking `doPartitialRead()` many times). Then these records are appended to those retrieved in #1.

3) return records to the user.

With such a scenario we do not need to have as many threads as consumers.